[Jello, Kakamotobi] 마을 및 우체통 그리기

Kakamotobi opened the pull request:
https://user-images.githubusercontent.com/79886384/229117034-de5e7352-e9bc-484e-8b8f-f6335dc8a8b8.mp4

### 주요 작업
- [x] **Mission 1**
  - [x] Draw villages on screen
    - [x] Choose random number of villages
    - [x] Create village
      - [x] Choose random size
      - [x] Choose random coordinates
    - [x] Render village
      - [x] Overlap check(10 times)
      - [x] Render or skip based on overlap check
  - [x] Draw mailboxes on screen
    - [x] Choose random number of mailboxes
- [x] **Mission 2**
  - [x] Button event
    - [x] Find all villages with a mailbox
      - [x] Create and use a custom `querySelector`
      - [x] Create and use a custom `querySelectorAll`
    - [x] Render village names in text view section
    - [x] Render village names in order of ascending size of mailbox
      - [x] Implement and use a sorting algorithm
    - [x] After 2 seconds, change the village border color to red
      - [x] Use `setTimeout` and `Promise` pattern to handle that delay

### 찾아본 키워드
- `Element.getBoundingClientRect()`
- `document.elementFromPoint()`
- `dataset` vs `getAttribute()`/`setAttribute()`
  - Decided to use `dataset` instead of `getAttribute`/`setAttribute` since it is the standard.
  - Also, there is no need to include the `data-*` prefix.
- Merge sort vs Radix sort
  - Decided to use Merge sort based on performance.
    - Merge sort TC - O(n log n)
      - Our case: O(8 * log8)
    - Radix sort TC - O(n * k)
      - Our case: O(8 * 6)

### 이슈
#### 남매 마을간 중첩
- Problem
  - Villages kept overlapping.
- Solution
  - Adjust the range of size and position of a village.
  - _Needed to change `every` to `some` when checking overlap between sibling villages._

Kakamotobi committed:
chore: Initialize configurations

- ESLint configurations
- Prettier configurations

Kakamotobi committed:
docs: Describe mission 1 feature requirements

Kakamotobi committed:
chore: Add fix ESLint on save setting

Kakamotobi committed:
chore: Setup base styles

- Structure markup.
- Add reset styles.
- Add CSS layouts.

Kakamotobi committed:
style: Add basic UI styles

- Check button
- Village box

Kakamotobi committed:
chore: Edit prettier config and cleanup

- Add `endOfLine: "auto"` to account for OS.
- Clean up duplicate prettier config in eslint.

Kakamotobi committed:
feat: Create VillageEl class

- VillageEl class represents the information to render a village DOM node.
- Contains methods to generate random size, position, and to check whether it overlaps with siblings.

Kakamotobi committed:
refactor: Change VillageEl name to VillageInfo

- Change `parent` to `parentEl`.
- Change `children` to `childrenEls`.

Kakamotobi committed:
refactor: Separate village info and DOM logic

- Separate village folder.
- Separate DOM logic from `VillageInfo` class.

Kakamotobi committed:
style: Cleanup layout

- Cleanup comments
- Apply `flex-grow` to `#canvas` instead of fixed width.

Kakamotobi committed:
chore: Change type to module

Kakamotobi committed:
fix: Incorrect overlap detection

- Adjust min/max values for generating random size and position.
- Change `every` to `some` in `isOverlappingWithSiblings`.
- Include `isValidSize()` method.
  - To check if the village passes the minimum size threshold.
- Separate `getRandomBetween` into utils.

Kakamotobi committed:
feat: Render villages

- Function to create valid village infos.
  - Validity is based on overlapping.
  - If village overlaps, regenerate random size and position.
    - Limit regeneration to 10 tries.
- Function to create village elements based on corresponding info.
- Function to append created village elements to the DOM.

Kakamotobi committed:
docs: Update features checklist and log dev issue

Kakamotobi committed:
fix: Prevent single child for canvas

- Edit condition for selecting a parent.
- Increase number of chances for regeneration due to overlap.
- Adjust proportional range of size and decrease minimum size.

Kakamotobi committed:
feat: Add mailbox to village element

- Decide whether to create mailbox in village element while creating the village elements.
- Add mailbox styles and image.

bakhacode commented on src/js/village/VillageInfo.js:
```diff
@@ -0,0 +1,102 @@
+    this.generatePosition();
+  }
+
+  generateSize() {
+    this.width = getRandomBetween(
+      this.parentInfo.width * 0.3,
+      this.parentInfo.width * 0.4
+    );
+    this.height = getRandomBetween(
+      this.parentInfo.height * 0.3,
+      this.parentInfo.height * 0.4
+    );
+  }
```
박하 - 0.3 0.4가 무엇을 의미하는지 알 수 있으면 좋을 것 같아요~! 🤓
27번도요!

hjsong333 replied:
static property로 sizeRatio 를 만들어서 매직넘버를 유의미한 이름으로 대체했습니다. 감사합니다! 😄

bakhacode commented on src/js/village/VillageInfo.js:
```diff
@@ -0,0 +1,102 @@
+    );
+  }
+
+  generatePosition() {
+    this.top = getRandomBetween(
+      (this.parentInfo.height - this.height) * 0.1,
+      (this.parentInfo.height - this.height) * 0.8
+    );
+    this.left = getRandomBetween(
+      (this.parentInfo.width - this.width) * 0.1,
+      (this.parentInfo.width - this.width) * 0.8
+    );
+
+    this.bottom = this.getBottom();
+    this.right = this.getRight();
+  }
+
```
박하 - getRandomBetween(top, left)를 줘서 하는 방식은 어떤가요? 함수내에서 값을 처리해서 받고 이후에 할당하는 동작만 있으면 깔끔해보일 것 같은데 어떤게 더 좋을까요 🧐 궁금합니다

hjsong333 replied:
코드로 충분히 의미가 전달된다고 생각해서 일단 그대로 두기로 정했습니다. 좋은 의견 감사합니다. 😄 👍

bakhacode commented on src/js/village/VillageInfo.js:
```diff
@@ -0,0 +1,102 @@
+    this.right = this.getRight();
+  }
+
+  getBottom() {
+    return this.top + this.height;
+  }
+
+  getRight() {
+    return this.left + this.width;
+  }
```
박하 - 변수에 바로 할당해도 될 것 같다고 보이는데 함수로 따로 뺀 이유가 있을까요?? 궁금궁금 😮

hjsong333 replied:
함수로 따로 뺀 이유는, 사용처가 굉장히 한정적이어서 굳이 속성값으로 가지고 다닐 필요가 없다고 생각했습니다.

hjsong333 replied:
논의 결과, 해당 함수들을 property로 변경하기로 결정했습니다.
메서드가 사용되는 곳은 제한적이지만, 해당 함수를 굉장히 빈번하게 호출하기 때문에 연산이 불필요하게 많아지기 때문입니다.

jsh3418 commented on src/js/village/VillageInfo.js:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  isOverlappingWithSiblings() {
+    if (this.parentInfo.childrenInfos.length === 0) return;
```
어떤 조건에서 early return이 되고 있는지 파악하기 어려운 것 같습니다.😭

hjsong333 replied:
noSibling 이라는 변수를 만들어서 대체해보겠습니다. 감사합니다. 😄

bakhacode commented on src/js/village/VillageInfo.js:
```diff
@@ -0,0 +1,102 @@
+    );
+  }
+
+  isValidSize() {
+    return this.width > 30 && this.height > 20;
+  }
```
토박 - 사이즈값을 생성할 때 넣지 않고 따로 함수로 뺀 이유가 궁금합니다

Kakamotobi replied:
저희 프로그램이 VillageInfo를 생성하고 조건에 만족하는 VillageInfo만 배열에 모으는 구조로 설계되었습니다.
새로 생성한 VillageInfo 객체가 valid한 size가 아닐 시에 배열에 추가하지 않게 하려고 했습니다. 
그래서 한 번 생성한 VillageInfo 객체는 스스로를 폐기할 수 없기 때문에 외부에서 size를 확인하는 방법이 필요해서 `isValidSize` 메소드를 만들었습니다.

Kakamotobi replied:
사이즈 값을 생성할 때 넣지 않은 이유는 부모 사이즈를 기준 비율로 생성하고 있는데 해당 함수는 우리가 생성하는 빌리지가 너무 작아지지 않게하기 위해 최소 사이즈를 가지도록 했습니다.
이 두 함수의 역할이 서로 충돌해서 의도치 않은 사이즈를 만들었고 이를 해소하기 위해 두 함수를 분리했습니다.

jjunyjjuny replied:
isValidSize로 최소한의 크기인지 먼저 확인하고,

isOverlappingWithSiblings > checkOverlap을 거쳐 생성된 마을이 유효한지 판단하는군요.
maxNumVillages가 20~30정도인 것은 버려지는 마을을 고려한 수치인가보네요..

만들어둔 유효성 검사 함수를 내부에서 활용해서 생성시에 조건이 만족되지 않았을때 generateSize/Position을 다시하는 방향은 고려해보셨을까요?

스스로를 폐기할 수 없지만 유효성 검사를 통과할때까지 스스로를 다시 init하는건 가능하지 않을까 싶습니다. 이러면 마을의 최소수를 보장할 수도 있을 거같아요. 폐기하는 방식은 최악의 케이스를 상정하면 0개의 마을이 생성될 수도 ..

hjsong333 replied:
피드백 주신 부분을 적용해서 내부에서 generateSize/Position을 다시하게 해봤습니다.
```JavaScript
constructor(name, parentInfo) {
  this.name = name;
  this.parentInfo = parentInfo;
  this.childrenInfos = [];
  this.generateSizeAndPosition();

  while (!this.isValidSize()) {
      this.generateSizeAndPosition();
    }
}
```
결과부터 말씀드리면 반복문 안에 갇혀서 페이지가 먹통이 되었습니다. 😢 

VillageInfo의 width, height을 계산할 때 **parent VillageInfo의 width, height의 30~40%**가 나오게 설정해뒀기 때문인 것으로 보입니다. width 50, height 50인 부모를 기반으로 생성되는 VillageInfo는 무한루프에 빠지게 되는 거죠.

이 문제를 피하려면 2가지 방식이 있을 것으로 생각했습니다. 
첫 번째로, 현재 구조에서는 (기존 방식처럼) 객체를 생성하는 곳에서 조건을 걸어서 생성을 막아줘야 할 것 같습니다. 
두 번째로, 클래스에서 자식 객체를 생성해주는 구조로 변경하면 유효한 크기일 때만 자식을 생성할 수 있을 것 같습니다.

jjunyjjuny replied:
30~40%의 수치나 최소 width/height의 조건이 있다면 자식을 만들 수 있는 최소한의 크기도 있겠군요.. 

조건을 더 추가하고 구조를 좀 변경하면 폐기가 아닌 다른 방식으로 구현할 수 있긴 하겠지만, 꼭 그래야만 하는 것은 아니니까요. 한 번 더 고민해봤다는 것에 의미를 두도록하시죠! 고생하셨습니다

bakhacode commented on src/js/village/VillageInfo.js:
```diff
@@ -0,0 +1,102 @@
+    name.className = "village-name";
+    name.textContent = this.name;
+
+    Object.assign(el.style, {
+      position: "absolute",
+      width: this.width + "px",
+      height: this.height + "px",
+      top: this.top + "px",
+      left: this.left + "px",
+    });
```
토박 - 크롱이 설명해줬을 때 찾았던 메서드인데 요렇게 쓰는군요 😗

Kakamotobi replied:
Template literal을 적용해서 더 간결하게 변경했습니다. 😄

jsh3418 commented on src/js/village/VillageInfo.js:
```diff
@@ -0,0 +1,102 @@
+    name.className = "village-name";
+    name.textContent = this.name;
+
+    Object.assign(el.style, {
+      position: "absolute",
+      width: this.width + "px",
+      height: this.height + "px",
+      top: this.top + "px",
+      left: this.left + "px",
+    });
```
한줄로 스타일을 지정할 수 있군요?!👍👍

bakhacode commented on src/js/village/VillageInfo.js:
```diff
@@ -0,0 +1,102 @@
+      left: this.left + "px",
+    });
+
+    innerWrapper.append(name, mailBox ?? "");
```
토박 - Nullish coalescing operator (??) 새로운 문법 배워갑니다. *96번줄에서 null을 반환하는게 확실해서 사용함

bakhacode commented on src/js/village/index.js:
```diff
@@ -0,0 +1,80 @@
+  childrenInfos: [],
+  width: canvasRect.width,
+  height: canvasRect.height,
+  top: 0,
+  left: 0,
```
토코 - 캔버스의 포지션이 relative이고 마을의 포지션이 absolute인데 top 이랑 left의 속성이 필요한지 궁금해요!

Kakamotobi replied:
좋은 지적 감사합니다. 덕분에 canvasInfo 객체를 점검하고 불필요한 속성들을 제거할 수 있었습니다. 👍

bakhacode commented on src/js/village/index.js:
```diff
@@ -0,0 +1,80 @@
+    if (checkOverlapWithSiblings(villageInfo)) {
+      villageNameCode++;
+      parentVillageInfos.push(villageInfo);
+      parentInfo.childrenInfos.push(villageInfo);
```
토박 - push가 마지막에 있어서 위에 있는 childrenInfos가 빈배열인데 반복하는 메서드가 있어서 의아했어요 😯

Kakamotobi replied:
언급하신 문제에 대해 고민을 해보니 개선하기 위해서는 Info 클래스의 상속 구조가 필요하다고 생각했습니다. 하지만 미션2에서 학습할 것들이 우선순위가 더 높다고 판단이 되어 리팩토링을 잠시 미루기로 결정했습니다. 😅

bakhacode commented on src/js/village/index.js:
```diff
@@ -0,0 +1,80 @@
+}
+
+function checkOverlapWithSiblings(villageInfo) {
+  for (let i = 0; i < 60; i++) {
```
60의 의미 궁금해요~
쿤디 - checkOverlapWithSiblings 함수가 check만 하고 함수로 읽혔는데 실패시에 재시도하는 다른 동작도 같이있는 것 같아서 분리하면 좀 더 동작을 읽기에 좋지 않을까용? 😤

Kakamotobi replied:
60은 village size와 position을 (sibling과 overlap할 시에) 재설정해줄 수 있는 기회의 숫자입니다.

Kakamotobi replied:
원래는 overlap할 때 size 및 position을 재설정해주려고 해서 의도하려던 함수의 역할이 맞았습니다.
하지만 함수는 하나의 역할 가지고 side effect를 배제하는 것이 의도치 않은 문제들을 방지할 수 있다는 점을 인지했습니다. 그래서 해당 함수의 분리를 시도해봤지만 쉽지않아서 최초의 의도를 다시 점검해봤습니다.
마을이 생성될 확률을 높이기 위한 기능이기 때문에 overlap할 때 size 및 position을 재설정해주는 과정이 반드시 필요하지는 않다는 결론을 냈습니다. Overlap할 시에 생성한 villageInfo 객체를 버리면 되기 때문입니다.

Kakamotobi committed:
fix: Applied PR review

- Clarified `if` conditions by using clearer variable names.
- Switched to template literal instead of string concatenation.
- Used static properties for size and position ratios.
- Set `bottom` and `right` properties instead of methods to compute them.
- Changed `break` to `continue` when the newly generated village size is not valid.
- Removed unnecessary properties on the `canvasInfo` object.
- Removed unnecessary logic for regenerating size and property upon overlap.

Kakamotobi committed:
refactor: Change "canvas" to "villageMap"

Kakamotobi committed:
docs: Update features checklist

Kakamotobi committed:
docs: Update features checklist for mission 2

Kakamotobi committed:
feat: Create customQuerySelector

- Receive and parse the selector and recursively traverse through the DOM to find the target element.

Kakamotobi committed:
docs: Update features checklist

hjsong333 committed:
feat: Create customQuerySelectorAll

- Receive and parse the selector and recursively traverse through the
  DOM to find all target elements.

hjsong333 committed:
feat: Render village names when click check button

- After click event, find all mailbox element.
- With the mailbox elements, grandparent is village.
- Render village names from attribute `data-name`.

Kakamotobi committed:
refactor: Clarify names and cleanup

- Use `dataset` instead of `setAttribute`/`getAttribute`.

Kakamotobi committed:
docs: Update features checklist

Kakamotobi committed:
feat: Render village names sorted by mailbox size

- Implement merge sort.
  - Tailor merge sort to accommodate value in object.
- Create `renderSortedVillageNamesBySize` function.
- Adjust markup to render sorted village names.
- Add `size` data attribute to village element.

Kakamotobi committed:
docs: Update features checklist and keywords

Kakamotobi committed:
refactor: Rename and restructure utils

- Create query selector factory function to reduce code duplication.
- Rename `mergeSort` to `customMergeSort`.

Kakamotobi committed:
feat: Highlight villages with a mailbox

- Create a function that delays a callback using promise and `setTimeout`.
- Create a function that highlights villages with a mailbox.

Kakamotobi committed:
docs: Update features checklist and keywords

Kakamotobi committed:
refactor: Cleanup unnecessary variable and code

jjunyjjuny commented on src/js/utils/customQuerySelector.js:
```diff
@@ -0,0 +1,87 @@
+  return [type, name];
+}
+
+function getSelectorType(selector) {
+  const selectorIdentifier = selector[0];
+  switch (selectorIdentifier) {
+    case ".":
+      return "class";
+    case "#":
+      return "id";
+    default:
+      return "tag";
+  }
+}
```
함수를 역할에 맞게 쪼개서 잘 만드신거 같아요!

지금도 미션을 해결하기 위한 용도로 충분하지만, 이왕 유틸을 만든다면 **사용법을 모르는 다른 사람이 사용한다** 가정하고 좀 더 안전하게 보완해 보는건 어떨까요?

switch의 default는 위에서 case로 처리하고 남은 모든 값이 동일한 상황이라고 확신하는 개발자의 의도가 느껴지는데 한 번쯤 정말 그럴까 의심해보는 것도 좋겠죠?

customQuerySelectorAll, customQuerySelector를 사용하는  사람이 일반적인 class, id, tag selector 외에 **다른 문자열을 넣는다면** 이 부분에서 type이 tag로 결정될거고, 이후 로직 어딘가에서 문제가 일어날 수도 있을 거 같아요 (아닐수도 있고요) 
예상해볼 수 있는 시나리오는 사용자가 실수로 id, class를 넣었는데 .이나 #을 깜빡하고 넣지 않았다면? 같은 .. (실제로 해봤던 실수..) 

물론, 좀 더 안전하게 만드나 지금을 유지하나 없는 셀렉터를 넣으면 null이 나오는건 기존의 querySelector와 동일할 거 같습니다만, 적어도 이 함수에서 잡아서 console이라도 찍어준다면 디버깅하기 좀 더 좋은 유틸이 되지 않을까 싶습니다.

어지간해서 오류없이 흘러가는 자바스크립트 특성상 오타처럼 사소한 문제가 발생했을 때 정작 어디가 문제인지 발견하는건 너무나 노가다거든요 .. 차라리 Error를 뿜어준다면 다행이지만…

Kakamotobi replied:
좋은 지적 감사합니다. 말씀하신대로 저도 깜빡하는 실수를 종종하는데 더 완성도 있는 유틸 함수를 위해 error같은 feedback을 주면 훨씬 user-friendly한 유틸이 될 것 같습니다.

jjunyjjuny commented on src/js/village/index.js:
```diff
@@ -0,0 +1,56 @@
+import { customQuerySelector } from "../utils/customQuerySelector.js";
+
+const villageMap = customQuerySelector("#village-map");
+const villageMapRect = villageMap.getBoundingClientRect();
```
처음보는 DOM API네요.. 역시 별게 다 있군용

hjsong333 replied:
저도 이번 미션을 진행하면서 처음 알게 됐습니다. 카카모토비의 검색 능력에 그저 감탄만...! 😲

Kakamotobi committed:
refactor: Detect invalid selector

- Add "tag" case for selector type in `customQuerySelector` for better feedback upon invalid selector.
